### ISSUE TITLE
Sudden horrible framerate

### ISSUE THREAD

**Issue by laserburn** [virtual-programming/specops-linux#5]
The game used to work without any issues for me, the only change I did is increase PoolSize setting to 512 which got rid of the stuttering. Then all of a sudden I start the game yesterday and it's completely unplayable. FPS in the main menu is about 3 frames per minute and it gets even worse once the game loads. Sound is choppy as well, like I'm trying to play this game on a 20 year old computer. Everything worked fine on previous sessions, game was running smoothly on max settings. I deleted and reinstalled the game, but no change.

My system specs gist:39d752dbbfab570b5f64
Content of the eon.txt gist:eb8664c37bedf7369edd

Not sure if I'm doing this right, first time on GitHub.

**Comment by licaon-kter** [comment]
After an uninstall did you delete this folder too, right?
```
$XDG_DATA_HOME/yager/specops/GameDocuments/My\ Games/SpecOps-TheLine/SRGame/Config/
```

**Comment by laserburn** [comment]
No, I didn't touch the configuration files. Tell me what you would like me to do.

**Comment by licaon-kter** [comment]
Since you said you modified the config, try to move it and rerun the game so it's created again in a clean slate, see what that behaves like.

**Comment by laserburn** [comment]
cd $XDG_DATA_HOME/yager/specops/GameDocuments/My\ Games/SpecOps-TheLine/SRGame/Config/
bash: cd: /yager/specops/GameDocuments/My Games/SpecOps-TheLine/SRGame/Config/: No such file or directory
But instead I now renamed ~/.local/share/yager folder and ran the game again. it seems to work fine, except that the stuttering is back again, but I know how to fix that. Would you like to see any of my old config files?

**Comment by licaon-kter** [comment]
Nope
Mark this as closed I guess.

**Comment by licaon-kter** [comment]
What happened, why re-open?

**Comment by laserburn** [comment]
Let me just add that it turns out the same thing happens every time I restart computer. Finished the game, so I won't be bothering with this any more, just to mention it in case others report the same problem.

**Comment by licaon-kter** [comment]
Well you could: delete the config, open game, close game, backup this config, restart computer, start game, see if it has issues and if it does, quit game and make a comparison between the newer config and the backup'ed one before the computer restarted. That would help.

**Comment by laserburn** [comment]
Sorry, I deleted it all. But I've left a link to this page on Steam support forum, some others claim to have the same problem. Perhaps this is also related to the "slow menu" issue that some are having.

**Comment by vpltd-peterm** [comment]
Might as well leave this open as it was not actually resolved. Without the information of what changed between configs, I couldn't speculate whats causing this

**Comment by robertc** [comment]
I have seen the 'really bad menu framerate' issue the last time I loaded it up, but I clicked on things and waited patiently and eventually the game started up and ran fine.  Is there anything specific I could do to investigate further?  Since I haven't played the game since I should be able to grab the eon.txt file when I get home later.

**Comment by robertc** [comment]
Here [is the eon.txt](https://gist.github.com/robertc/6d9019f26052f15de79e).  I was playing again tonight, the menu was really laggy on the first loop though the theme song, then became a lot more responsive on the second loop through.  As in, as the tune ended and started again that's when it became more responsive.

**Comment by manwesulimo2004** [comment]
I also just encountered this issue. Game was working fine a couple days ago. What kind of info should I provide to help get this fixed?

**Comment by licaon-kter** [comment]
Try to backup your config folder (where ever it is):

```
$XDG_DATA_HOME/yager/specops/GameDocuments/My\ Games/SpecOps-TheLine/SRGame/Config/
```

Restart computer, start game, test, if it's ok, compare the new Config folder with your backup, using *diff* or whatever.

**Comment by manwesulimo2004** [comment]
So after removing the old config folder and firing up the game again, everything seems to be working at first glance (with the really low default settings). The three config files that have changed are ```SREngine.ini```, ```SRGame.ini``` and ```SRInput.ini```.

1. I presume the changes in ```SRInput.ini``` don't matter (keybinding, etc.).
2. The only change in ```SRGame.ini``` was the ```[DefaultPlayer]``` ```team``` value (probably meaningless).
3. The before and after for ```SREngine.ini``` can be viewed in [this gist](https://gist.github.com/manwesulimo2004/4ff47411eade5432e228).

I have also appended the ```eon.txt``` for working and broken configs in the gist linked in point 3. It might also be worth mentioning the ```eONprecompiledShaders32.dat``` was about 12.4 MB on the broken config and about 840 KB for the working config. I don't know exactly what this file is for but that may simply be due to the fact I haven't actually left the main menu since creating the new config.

**Comment by vpltd-peterm** [comment]
eONprecompiledShaders32.dat is a cache file of all the shaders the game has come across, converted to GLSL and compiled. 

If you have the old copy, try putting it back... i'd be interested to know if the frame rate problems return when you do!

**Comment by manwesulimo2004** [comment]
You were on to something there. I tried the old file and it did re-introduce the framerate issue. Interestingly it wasn't fixed after restoring the newer file though. Instead I had to remove the file completely to get things working again.

So whatever is going on here seems to be "re-corrupting" that file after certain play sessions.

**Comment by rkfg** [comment]
Just had this as well. However, my setup isn't quite standard, I use pf-kernel (compiled myself) and NVIDIA driver from geforce.com, i.e. not from the repository. All of this is on Debian testing amd64 and the desktop is managed by Awesome WM. I've never got any troubles because of this, games run slightly better, sound isn't choppy where it is on the stock kernel and so on.

However, today I had several games not launching at all (silently exiting) and SpecOps behaving weird, like already described in this thread. Those games were fixed after reinstalling the videodriver, probably because I've updated the kernel yesterday and haven't run any games after that so the problem only surfaced today. But SpecOps was freezing even after that. My guess is that it's recompiling shaders but doing it for too long for whatever reason. I've already completed the game and had basically zero issues except them freezes on loading here and there but that's common.

The very distinctive symptom of the bug is the slow intro videos. When the Steam FPS indicator is enabled it shows 20 FPS instead of 30 and everything is 1.5 times slower (this should not happen normally but it happens). Clearly visible on the Unreal logo sequence, it's like in slow-mo, not choppy. In the main menu I get a slideshow. However, if I leave it like this for some time it becomes good again eventually, 60 FPS and such. After restarting the game loads just fine, intro is fast and smooth.

For the record, my eONprecompiledShaders32.dat is 22461436 bytes. Renaming the entire ~/.local/share/yager directory doesn't introduce the bug. There's also another cache in ~/.nv/GLCache, clearing it also doesn't break anything.

**Comment by rkfg** [comment]
Hmm, rebooted the PC and it's back. It's like 5 seconds between frames, if I move the mouse the response happens in 5 seconds. The game uses 230% of CPU and in about 4 minutes after launching becomes fine. CPU load drops to 155-170%.

**Comment by rkfg** [comment]
Renaming that file, rebooting and starting the game helps, as @manwesulimo2004 said. Probably the cache becomes too bloated for whatever reason. I've [uploaded](http://rkfg.me/tmp/eONprecompiledShaders32.zip) the broken cache to my server, feel free to analyze it.

**Comment by manwesulimo2004** [comment]
If it helps, I also have an nvidia card. I use the standard repo drivers (currently 331.113-0ubuntu0.0.4) and the default kernel though (3.13.0-24-generic on Linux Mint 17.0).

**Comment by rkfg** [comment]
And by the way, it constantly writes to disk @ about 400 Kb/sec according to iotop:
```
 3616 be/4 btruser     0.00 B/s  401.07 K/s  0.00 %  0.00 % ./specops.i386 [E.d3d9.OpenGL]
```

Probably it does it in the render thread and that's why it's so slow until it finishes building the shader cache. With a little help of procfs I've managed to find out what it writes exactly and it's ~/.nv/GLCache/e51b2a21cdb3b205dc9f1bf2937745f2/a5d631c3fb538d5e/53ded2211899847a.bin and .toc file in the same directory. The resulting .bin file is 118611647 bytes long, wow. Removing the entire GLCache dir while leaving eONprecompiledShaders32.dat triggers the bug, actually. No need to reboot even.

**Comment by vpltd-peterm** [comment]
OK so this has narrowed it down to shaders, and even possibly to a problem in the nvidia driver. We will take a look at this when we can!

**Comment by licaon-kter** [comment]
I'm on the beta nVidia driver and I don't see this, can you ty it: https://devtalk.nvidia.com/default/topic/833031/unix-graphics-announcements-and-news/-linux-solaris-and-freebsd-driver-352-09-beta-/

**Comment by rkfg** [comment]
Tried with 352.09, cleared .nv and everything is the same. It's not the driver issue probably, it's the bloated shader cache that's broken. Or do you mean it recognizes the created cache after reboot and doesn't dump it again?

**Comment by licaon-kter** [comment]
Can you upload Config folder and shader cache somewhere so I can replace mine and test?

**Comment by rkfg** [comment]
[I've already done this](https://github.com/virtual-programming/specops-linux/issues/5#issuecomment-105647476), download it and try for yourself.

**Comment by licaon-kter** [comment]
and Config

**Comment by rkfg** [comment]
It's pretty default but if it makes any difference, [here, have it](http://rkfg.me/tmp/confs.zip).

**Comment by licaon-kter** [comment]
It's modified on game run hence maybe something is setup different. 10x.

**Comment by licaon-kter** [comment]
Retested a bit, with your shader file, game starts with low fps, this file is recreated and grows

```
~/.nv/GLCache/8ed6def4e3c4f9af4d971762002c666d/512f7de3a35d6a6d/5852cdcc5d6a16b7.bin
```
and while it grows, it talkes a minute or two, the fps stays low, when it gets to ~ 110Mb then the fps rises slowly up to 60 ( vsync on ), and stays there.

Now, I see that your shader file is ~21Mb, how many levels/much content did you play?

Since my fps starts up at 60, I tried to build up my shader file thinking that maybe the problem is it's size, loaded every level up to 13 ( yeah that's where I am currently in game ), and roamed a bit as much as I was allowed by the gameplay, and my shader file got up to ~12Mb only: http://is.gd/RGt1UF

**Comment by rkfg** [comment]
That's the exact behavior I have. I've completed the game and ran one game in multiplayer, that's all. After a couple of days I tried to run it again and it was like this. There are two issues here as I see it:

1. The game should load/unpack/compile/whatever it does not in parallel with menu but somehow before it starts interacting with the player in any way, be it the main menu, intro or anything else. A progress bar would be just fine but it would require modifying the game code which, if I understand correctly, isn't how eON works. They may make a launcher program that explicitly loads the cache so it populates the driver's cache and only then launch the actual game binary.
1. For a reason unknown this driver's cache doesn't persist across reboots for the game. The file is there, yes, but it's getting rewritten after reboot. Dying Light also has a file there (/f8e174f44f867fde485c869d1f3fed0d/a5d631c3fb538d5e/5ca51d6bdeeae8af.bin and .toc) but the game loads quickly enough and those file are tiny, 2.2 Mb only. Maybe it's not a bug that the file invalidates after reboot. Such a graphics-intensive game as Dying Light doesn't have a big dump there so there's no problems.

**Comment by dubigrasu** [comment]
@rkfg 
"For a reason unknown this driver's cache doesn't persist across reboots for the game."

From what I've seen so far that's the standard behavior for every game or application.
http://us.download.nvidia.com/XFree86/Linux-x86/346.72/README/openglenvvariables.html

**Comment by rkfg** [comment]
The document you're linking describes the opposite:

>Caches are **persistent across runs of an application.** Cached shader binaries are specific to each driver version; **changing driver versions will cause binaries to be recompiled.**

It doesn't say anything about simple reboots, so the recompiling should occur only on driver update or downgrade.

**Comment by dubigrasu** [comment]
@rkfg 
Can't say how this actually works since I haven't found any documentation except the one I posted as reference.
What I've meant is that those cache files are only temporary and they will be cleared eventually, although I'm not sure what exactly triggers that.
It could be the age of the cache or after the .nv folder reaches a certain size, or some other criteria.

**Comment by rkfg** [comment]
@dubigrasu size could be a reason to clean it up. Other games don't make such large caches as this one. Either way eON probably tries to speed things up by compiling all the shaders it knows about as early as it can and the outcome is what we have.

**Comment by vpltd-peterm** [comment]
the eON precompiled shader file is storing the GLSL versions of whatever HLSL/DirectX shader asm is sent from the game. We don't cache compiled GLSL as there is no real sense for us to do it - both nvidia and amd's drivers cache this already.

The problem with SpecOps (or Unreal Engine) is it sends new shaders as it uses them, and they are compiled on the fly. This is because on DirectX, this is fast. On GL, compiling GLSL is slower and also stalls the pipeline. We try to minimise its effects, but it is one of the flaws of OpenGL that holds back performance in comparison.

**Comment by licaon-kter** [comment]
"The problem with SpecOps (or Unreal Engine) is it sends new shaders as it uses them, and they are compiled on the fly."

umm why does it need ALL the shaders at the first menu where it says "press any key to continue" then?

**Comment by vpltd-peterm** [comment]
I am referring to as you play through the game. The menu is still a rendered scene and has shaders. The first time these shaders are handed to eON, they are converted to GLSL, stored in our precompiled shader file (maybe this is badly named and should be eONConvertedShaders instead). 

When the game tries to load a shader thats already been converted to GLSL, it is matched via hashing, and the relatively expensive shader -> GLSL stage is skipped. However, the GL library still has to process the GLSL shader, and this causes a delay.

However, it shouldnt cause the issue this thread is about... we didn't see it happening during development. It's on the to-do list.

**Comment by steamlinux** [comment]
I have low fps now
if I understand, I need this
```rm -f  $XDG_DATA_HOME/yager/specops/eONprecompiledShaders32.dat```
before I have 4 fps, after remove this I have 2-3 fps
sorry of my bad english

**Comment by steamlinux** [comment]
and I try your eONprecompiledShaders32.dat, nothing
after I remove eONprecompiledShaders32.dat and have 60 fps
thank's for this thread
